vertical-pod-autoscaler/pkg/recommender/input/controller_fetcher/controller_fetcher.go

```diff
@@ -82,10 +83,44 @@ type controllerFetcher struct {
 	scaleNamespacer scale.ScalesGetter
 	mapper          apimeta.RESTMapper
 	informersMap    map[wellKnownController]cache.SharedIndexInformer
+	controllerCache controllerCacheStorage
+}
+
+func (f *controllerFetcher) periodicallyRemoveExpiredFromCache(ctx context.Context, period time.Duration) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
+			f.controllerCache.RemoveExpired()
+		}
+	}
+}
+
+func (f *controllerFetcher) periodicallyRefreshCache(ctx context.Context, period time.Duration) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
+			keysToReferesh := f.controllerCache.GetKeysToRefresh()
+			klog.Info("Starting to refresh entries in controllerFetchers controllerCache")
+			for _, item := range keysToReferesh {
+				scale, err := f.scaleNamespacer.Scales(item.namespace).Get(context.TODO(), item.groupResource, item.name, metav1.GetOptions{})
+				f.controllerCache.Refresh(item.namespace, item.groupResource, item.name, scale, err)
+			}
+			klog.Infof("Finished refreshing %d entries in controllerFetchers controllerCache", len(keysToReferesh))
+		}
+	}
+}
+
+func (f *controllerFetcher) Start(ctx context.Context, removePeriod, refreshPeriod time.Duration) {
```

Comment:
Could we actually for simplicity have one cache refresh function that would delete items that should be deleted and refresh the ones that are stale but still valid?

Reply:
Done.